### PR TITLE
MM-53478 : Improve design of keyword that trigger notification

### PR DESCRIPTION
**Pull request by M-ZubairAhmed**
#### Summary
- General code improvements around keywords that trigger notifications in `user_settings_notifications.tsx`
- Used react-select/creatable for the multi-input tag design

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-53478

#### Screenshots
| Before 	| After 	|
|--------	|-------	|
| image        	| image       	|

#### Release Note

```release-note
NONE
```

### CODE REVIEW

**Comment by M-ZubairAhmed** [comment]
@abhijit-singh Few points
- I have not added the 'comma' press to create tab, the reason for that being is pressing `tab` to create tags is common user behaviour eg. Twitter hashtags. But comma could be added, but I need to handle that separately, adding that I thought isn't worth the effort. But let me know.
- Design tokens are off, as you have mentioned the design for that was not finalized, so I pushed it as per your suggestion. If some changes need to be made, do let me know.

**Review by abhijit-singh — CHANGES_REQUESTED** [review]
Thanks @M-ZubairAhmed, looks good! Just one request — If the user types in a single keyword, but doesn't hit TAB and instead clicks on `Save`, can we make sure that keyword gets saved and the user sees a 'chip' with it when they come back to this UI? Currently it gets lost of you don't press TAB.

**Comment by M-ZubairAhmed** [comment]
> Thanks @M-ZubairAhmed, looks good! Just one request — If the user types in a single keyword, but doesn't hit TAB and instead clicks on `Save`, can we make sure that keyword gets saved and the user sees a 'chip' with it when they come back to this UI? Currently it gets lost of you don't press TAB.

I was checking this but currently the only time the values are returned from this component (which is a third party component) are when they are actually selected. This component is actually a createable select with the menu option hidden :P So in conclusion its a lot of effort to reach the internal state of the third party component and get it overriding the onchange.

**Comment by abhijit-singh** [comment]
Okay. I think this is a significant issue. From the user's perspective, it is a very valid flow to enter a keyword and click on `Save`. 

Is there something else we can do maybe? Throw an error asking the user to press `TAB` if they want the keyword to get saved is one other approach.

cc @johndavidlugtu

**Comment by johndavidlugtu** [comment]
> I was checking this but currently the only time the values are returned from this component (which is a third party component) are when they are actually selected. This component is actually a createable select with the menu option hidden :P So in conclusion its a lot of effort to reach the internal state of the third party component and get it overriding the onchange.

Thanks for the context @M-ZubairAhmed -- I may need your help understanding a bit better. Does this mean that if a user clicks save, we can't effectively "press tab for the user"? Basically, if a keyword isn't currently in a pill, we put that keyword into the pill.

In the current experience, adding a keyword (without including a comma) does save the keyword.

In any case, I'm open to the idea of alternative solutions if the ideal experience above is a lot of effort more than it's worth.
cc: @abhijit-singh

**Comment by M-ZubairAhmed** [comment]
> > I was checking this but currently the only time the values are returned from this component (which is a third party component) are when they are actually selected. This component is actually a createable select with the menu option hidden :P So in conclusion its a lot of effort to reach the internal state of the third party component and get it overriding the onchange.
> 
> Thanks for the context @M-ZubairAhmed -- I may need your help understanding a bit better. Does this mean that if a user clicks save, we can't effectively "press tab for the user"? Basically, if a keyword isn't currently in a pill, we put that keyword into the pill.
> 
> In the current experience, adding a keyword (without including a comma) does save the keyword.
> 
> In any case, I'm open to the idea of alternative solutions if the ideal experience above is a lot of effort more than it's worth. cc: @abhijit-singh

@johndavidlugtu I have added the video let me know if this helps in understanding.
- type in the custom keywords field
- press tab, and it will create a pill around the keyword
- type another keyword, don't press tab and just press save
- keyword will not be created because it was not registered in the first place (without hitting tab to create)

https://github.com/mattermost/mattermost/assets/17708702/1ccb24f5-c0c4-4e33-bbd8-25d69435f205

**Comment by johndavidlugtu** [comment]
@M-ZubairAhmed understood that part. My question is, can we "press tab on behalf of the user" if we detect a keyword that hasn't been registered? or something similar that would achieve that?

**Comment by M-ZubairAhmed** [comment]
> Thanks @M-ZubairAhmed, looks good! Just one request — If the user types in a single keyword, but doesn't hit TAB and instead clicks on `Save`, can we make sure that keyword gets saved and the user sees a 'chip' with it when they come back to this UI? Currently it gets lost of you don't press TAB.

I handled this case by creating a chip on the blur of the input instead

**Review by abhijit-singh — APPROVED** [review]
Looks good, thanks @M-ZubairAhmed

File: webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx [hunk 1]
```diff
@@ -79,86 +84,90 @@ function getNotificationsStateFromProps(props: Props): State {
...
+            usernameKey = mentionKeys.includes(props.user.username);
 
-            customKeys = keys.join(',');
+            // Check if there are any keys in the list, if so, set the checkbox of custom keys to true
+            isCustomKeysWithNotificationInputChecked = customKeysWithNotification.length > 0;
         }
 
-        if (user.notify_props.first_name) {
-            firstNameKey = user.notify_props.first_name === 'true';
+        if (props.user.notify_props.first_name) {
```
Comment (larkox) [inline webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx:165]: I don't think the `if` are needed (nor the next one). If it is not set, it is undefined, and undefined is also distinct from 'true'.

File: webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx [hunk 2]
```diff
@@ -223,45 +230,42 @@ export default class NotificationsTab extends React.PureComponent<Props, State>
         data.comments = this.state.notifyCommentsLevel;
         data.auto_responder_active = this.state.autoResponderActive.toString() as UserNotifyProps['auto_responder_active'];
         data.auto_responder_message = this.state.autoResponderMessage;
+        data.first_name = this.state.firstNameKey.toString() as UserNotifyProps['first_name'];
+        data.channel = this.state.channelKey.toString() as UserNotifyProps['channel'];
```
Comment (larkox) [inline webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx:234]: Why are the castings needed?

File: webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx [hunk 3]
```diff
@@ -223,45 +230,42 @@ export default class NotificationsTab extends React.PureComponent<Props, State>
...
-                if (result) {
-                    this.handleUpdateSection('');
-                    this.setState(getNotificationsStateFromProps(this.props));
-                } else if (err) {
-                    this.setState({serverError: err.message, isSaving: false});
-                }
-            });
+        stopTryNotificationRing();
+
+        const {data: updatedUser, error} = await this.props.updateMe({notify_props: data}) as ActionResult<Partial<UserProfile>, ServerError>;
```
Comment (larkox) [inline webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx:257]: Why the casting is needed?

**Review by larkox — APPROVED** [review]
LGTM, just some nits and questions.

**Review by yasserfaraazkhan — CHANGES_REQUESTED** [review]
Hi @M-ZubairAhmed , this is working as expected. 
words are getting highlights, @ mention section is population correctly, Editing a message to add keywords is also working fine.


- couple of e2e test might need fixing
- Should user save a sentence ?
**in figma**

Screenshot 2023-07-31 at 4 21 46 PM

**in the spinwick**

Screenshot 2023-07-31 at 4 22 30 PM

**Comment by M-ZubairAhmed** [comment]
> * Should user save a sentence ?

No not in this PR

> * couple of e2e test might need fixing

Let me fix those

File: webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx (hunk 2 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx:234]: yes removed the casting

File: webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx (hunk 1 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx:165]: yes correct removed it

File: webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx (hunk 3 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx:257]: this will be fixed in MM-46907, in very future :P

**Comment by harshilsharma63** [comment]
So some issues I see in testing-

1. When creating more than one keywords with exact same text, we get a duplicate react key error-
```
ncountered two children with the same key, `hello`. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version.
```

https://github.com/mattermost/mattermost/assets/18575143/91fc28a3-e746-46b7-96c7-f9f3fb9c747d

2. When creating a keyword with comma, it looks different while creating and after saving and re-opening the modal


https://github.com/mattermost/mattermost/assets/18575143/cc080b0c-41f3-4709-9b40-543b188d2342

**Comment by M-ZubairAhmed** [comment]
> When creating a keyword with comma, it looks different while creating and after saving and re-opening the modal

@abhijit-singh regarding this I don't think we can accommodate server changes for this PR. Can we revisit this case with server support for commas in the subsequent work on spaces?

**Comment by abhijit-singh** [comment]
> > When creating a keyword with comma, it looks different while creating and after saving and re-opening the modal
> 
> @abhijit-singh regarding this I don't think we can accommodate server changes for this PR. Can we revisit this case with server support for commas in the subsequent work on spaces?

I don't think we can go ahead with this because it is functionally incorrect if the user adds any commas in their keywords. Maybe we just don't merge this until the server changes are in? 

I believe the priority was to do `Highlighting keywords without notifications` and these design updates can wait. 
cc @johndavidlugtu

**Review by harshilsharma63 — CHANGES_REQUESTED** [review]
Pending on comma changes.

File: webapp/channels/src/i18n/en.json [hunk 4]
```diff
@@ -5421,27 +5421,26 @@
...
   "user.settings.notifications.never": "Never",
-  "user.settings.notifications.noWords": "No words configured",
   "user.settings.notifications.off": "Off",
   "user.settings.notifications.on": "On",
   "user.settings.notifications.onlyMentions": "Only for mentions and direct messages",
   "user.settings.notifications.push": "Mobile Push Notifications",
   "user.settings.notifications.push_notification.status": "Trigger push notifications when",
   "user.settings.notifications.push_threads": "When enabled, any reply to a thread you're following will send a mobile push notification.",
   "user.settings.notifications.push_threads.allActivity": "Notify me about threads I'm following",
+  "user.settings.notifications.sensitiveCustomWords": "Other non case-sensitive words, press TAB to seperate keywords:",
```
Comment (cwarnermm) [inline webapp/channels/src/i18n/en.json:5434]:
```suggestion
  "user.settings.notifications.sensitiveCustomWords": "Other non case-sensitive words, press TAB to separate keywords:",
```

File: webapp/channels/src/i18n/en.json [hunk 5]
```diff
@@ -5421,27 +5421,26 @@
   "user.settings.notifications.header": "Notifications",
   "user.settings.notifications.icon": "Notification Settings Icon",
   "user.settings.notifications.info": "Desktop notifications are available on Edge, Firefox, Safari, Chrome and Mattermost Desktop Apps.",
-  "user.settings.notifications.mentionsInfo": "Mentions trigger when someone sends a message that includes your username (\"@{username}\") or any of the options selected above.",
+  "user.settings.notifications.keywordsWithNotification.extraInfo": "Notifications get triggered when someone sends a message that includes your username (\"@{username}\") or any of the options selected above.",
```
Comment (cwarnermm) [inline webapp/channels/src/i18n/en.json:5424]:
```suggestion
  "user.settings.notifications.keywordsWithNotification.extraInfo": "Notifications are triggered when someone sends a message that includes your username (\"@{username}\") or any of the options selected above.",
```

**Comment by cwarnermm** [comment]
@M-ZubairAhmed - I've provided suggestions to consider for the product strings.

File: webapp/channels/src/i18n/en.json (hunk 4 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/i18n/en.json]: Done ✅

File: webapp/channels/src/i18n/en.json (hunk 5 above)
Comment (M-ZubairAhmed) [reply webapp/channels/src/i18n/en.json]: Done ✅

**Review by cwarnermm — APPROVED** [review]

**Review by harshilsharma63 — APPROVED** [review]
Looks good!

**Review by yasserfaraazkhan — APPROVED** [review]

**Pull request merged**